### PR TITLE
[Structural] Generic linear elastic law

### PR DESCRIPTION
**Pull request by rubenzorrilla**
This simple constitutive law allows the user to compute the material response by saving a custom C tensor in the `ELASTICITY_TENSOR` variable. Tests have been added accordingly.

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_constitutive/user_provided_linear_elastic_law.cpp [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+template<unsigned int TDim>
+void UserProvidedLinearElasticLaw<TDim>::CalculatePK2Stress(
+    const Vector& rStrainVector,
+    Vector& rStressVector,
+    ConstitutiveLaw::Parameters& rValues
+    )
+{
+    const Properties& r_material_properties = rValues.GetMaterialProperties();
+    const Matrix C = r_material_properties[ELASTICITY_TENSOR];
+    rStressVector = prod(C, rStrainVector);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/user_provided_linear_elastic_law.cpp]:
```suggestion
    noalias(rStressVector) = prod(C, rStrainVector);
```

File: applications/StructuralMechanicsApplication/custom_constitutive/user_provided_linear_elastic_law.cpp [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<unsigned int TDim>
+void UserProvidedLinearElasticLaw<TDim>::CalculateCauchyGreenStrain(
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/user_provided_linear_elastic_law.cpp]: This is duplicated in the CL utilities

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/user_provided_linear_elastic_law.cpp]: https://github.com/KratosMultiphysics/Kratos/blob/7d87c30d57054bed668e0a6ebb78346792d5a350/applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp#L373

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Some comments. (May ask an example of use of this?)

File: applications/StructuralMechanicsApplication/custom_constitutive/user_provided_linear_elastic_law.cpp (hunk 2 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/user_provided_linear_elastic_law.cpp]: My blocking reason is this, only this

**Review by philbucher — CHANGES_REQUESTED** [review]
please move this law to the advanced_constitutive laws

otherwise looks good to me :+1:

**Comment by rubenzorrilla** [comment]
> please move this law to the advanced_constitutive laws
> 
> otherwise looks good to me +1

Actually we intentionally place this one here. Note that this one makes it possible to use any constitutive model (of course linearised) provided that you have the C tensor.

**Comment by rubenzorrilla** [comment]
> Some comments. (May ask an example of use of this?)

Someone gives you a C tensor but you have no constitutive law. We have been using it to solve anisotropic materials.

File: applications/StructuralMechanicsApplication/custom_constitutive/user_provided_linear_elastic_law.cpp (hunk 2 above)
Comment (rubenzorrilla) [reply applications/StructuralMechanicsApplication/custom_constitutive/user_provided_linear_elastic_law.cpp]: > This is duplicated in the CL utilities

Cool. I didn't know about this.

File: applications/StructuralMechanicsApplication/custom_constitutive/user_provided_linear_elastic_law.cpp (hunk 1 above)
Comment (rubenzorrilla) [reply applications/StructuralMechanicsApplication/custom_constitutive/user_provided_linear_elastic_law.cpp]: Thanks!

**Comment by loumalouomega** [comment]
> 
> 
> > Some comments. (May ask an example of use of this?)
> 
> Someone gives you a C tensor but you have no constitutive law. We have been using it to solve anisotropic materials.

But it converges well?, if the tensor is not the derivative of the stress should lose the quadratic convergence. Just asking. Is not better to define a consistent anisotropic law?

**Comment by loumalouomega** [comment]
Travis fails BTW

File: applications/StructuralMechanicsApplication/custom_constitutive/user_provided_linear_elastic_law.cpp [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+/***********************************************************************************/
+
+template<unsigned int TDim>
+void UserProvidedLinearElasticLaw<TDim>::CalculateElasticMatrix(
+    Matrix& rConstitutiveMatrix,
+    ConstitutiveLaw::Parameters& rValues
+    )
+{
+    const Properties& r_material_properties = rValues.GetMaterialProperties();
+    rConstitutiveMatrix = r_material_properties[ELASTICITY_TENSOR];
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/user_provided_linear_elastic_law.cpp]:
```suggestion
    noalias(rConstitutiveMatrix) = r_material_properties[ELASTICITY_TENSOR];
```
I forgot this one too

**Comment by rubenzorrilla** [comment]
> > > Some comments. (May ask an example of use of this?)
> > 
> > 
> > Someone gives you a C tensor but you have no constitutive law. We have been using it to solve anisotropic materials.
> 
> But it converges well?, if the tensor is not the derivative of the stress should lose the quadratic convergence. Just asking. Is not better to define a consistent anisotropic law?

It is linear so it should. And in any case, you may face a situation in which someone provides you with the C tensor but you don't want or you can't (it is confidential) implement the complete constitutive model.

**Review by loumalouomega — DISMISSED** [review]
Removing my block

**Review by philbucher — APPROVED** [review]
approving since this most probably will also go to the core eventually

**Review by RiccardoRossi — APPROVED** [review]
+1

**Comment by loumalouomega** [comment]
I am not sure about that. If S=C:E then it should

**Pull request merged**